standards/docker_standards.md

```diff
@@ -0,0 +1,230 @@
+  ```
+  ARG PORT=3000
+  ENV PORT ${PORT}
+  ```
+
+#### .Net Core
+- a parent image should exist for both the .Net Core SDK and the .Net Core Runtime
+- extend minimal .Net Core base image
+- set a fixed major version of the base image
+- set the `ASPNETCORE_ENVIRONMENT` environment variable to production in parent
+- run as the `www-data` user and not `root`
+- ensure the working directory is owned by the www-data user - e.g. /app
+- a remote debugger should be added to the SDK file to aid development and testing
+  ```
+  RUN apt-get update \
+  && apt-get install -y --no-install-recommends unzip \
+  && curl -sSL https://aka.ms/getvsdbgsh | bash /dev/stdin -v latest -l /vsdbg
+  ```
+
+## Container scanning
+Container scanning tools collate vulnerability reports from a number of sources to determine security issues within operating system kernels, libraries, and software packages.
+
+Containers should be regularly scanned using tools such as Clair, Anchore Engine, Aqua Trivy or JFrog XRay.
+
+A comparison of the different tools is available in [Confluence](https://eaflood.atlassian.net/wiki/spaces/FPS/pages/1408172407/Container+Image+Scanning+Tools)
+
+## Multi stage builds
+Dockerfiles should implement multi stage builds to allow different container configuration and behaviour to be created for development, test and production.
+
+Below is an example multi stage build which is intended to use the Future Farming and Countryside (FFC) Node.js parent image.
+
+```
+# Base stage installs production dependencies
+ARG REGISTRY=562955126301.dkr.ecr.eu-west-2.amazonaws.com
+ARG BASE_VERSION=1.0.0
+ARG DEV_VERSION=1.0.0
+FROM $REGISTRY/ffc-node-base:$BASE_VERSION as base
+ARG PORT=3000
+ENV PORT ${PORT}
+USER node
+WORKDIR /home/node
+COPY --chown=node:node package*.json ./
+RUN npm ci
+
+# Development stage installs devDependencies, builds app from source and declares a file watcher as the default command
+FROM $REGISTRY/ffc-node-development:$DEV_VERSION AS development
+EXPOSE ${PORT} 9229 9230
+USER node
+WORKDIR /home/node
+COPY --from=base --chown=node:node /home/node/package*.json ./
+COPY --from=base --chown=node:node /home/node/node_modules ./node_modules
+RUN npm install
+COPY --chown=node:node app/ ./app/
+RUN npm run build
+CMD [ "npm", "run", "start:watch" ]
+
+# Test stage copied in Jest configuration and declares the test task as the default command
+FROM development AS test
+USER node
+WORKDIR /home/node
+COPY --chown=node:node jest.config.js ./jest.config.js
+COPY --chown=node:node test/ ./test/
+CMD [ "npm", "run", "test" ]
+
+# Production stage exposes service port, copies in built app code and declares the Node app as the default command
+FROM base AS production
+USER node
+WORKDIR /home/node
+EXPOSE ${PORT}
+COPY --from=development /home/node/app/ ./app/
+CMD [ "node", "app/index" ]
+```
+
+## Docker Compose
+Docker compose can be used for local development and within a CI/CD pipeline to build docker containers.
```

Comment:
I'd like to see some guidance or ideally a standard on how/where to structure the `docker-compose.yml` files in a multi-repo project.

Reply:
In FFC for our demo service, we've created a development repo which clones all others repos and runs all of their relevant docker-compose files in one command to bring the end to end service up in one docker network.  It also creates single instances of shared dependencies such as message queues.  I see the Ivory project have done something similar.

For scenarios where there are only one or two other repos then it maybe unnecessary effort to create a development repo.

Now that's local development of course.  When it comes to CI and deploying multiple repos alongside each other for end to end/integration testing or in support of a PR build then that's different ball game and will be very subjective to specific test strategies, programme needs, service complexity so wouldn't really be keen on trying to set a standard.

I'll document the approach FFC and Ivory have taken and see what opinion are.

Reply:
I've added some guidance on how to achieve this in a new section, "Composing multiple repositories for local development"